### DISCUSSION TITLE
Launching Rocket without codegen

### DISCUSSION THREAD

**Discussion by bubelov** [rwf2/Rocket#1772] [Questions] [answered]
Are there any examples of manually starting a server, without code generators? I'd like to make an unified binary which accepts and parses command line arguments before deciding what to do. Starting a server is only one of the possible actions, so I need full control over the main function in order to figure out if user wants to start a server or do something else

**Comment by baptiste0928** [comment, marked as answer]
You can manually launch a Rocket instance with the [`launch()`](https://api.rocket.rs/v0.5-rc/rocket/struct.Rocket.html#method.launch) method :

```rs
#[rocket::main]
async fn main() {
    rocket::build()
        .mount("/hello", routes![world])
        .launch()
        .await;
}
```

More information : https://rocket.rs/v0.5-rc/guide/overview/#launching

**Reply by bubelov** [reply]
@baptiste0928 yes, I read the doc before asking questions. My problem with this approach is the fact that it still uses rocket-depended codegen annotation (`#[rocket::main]`). What I'm trying to do is to handle CLI args first and then decide whether I need to initialize async runtime and start a Rocket server

**Reply by baptiste0928** [reply]
According to the documentation, `#[rocket::main]` launches a custom tokio runtime. It seems to be launched by the internal [`async_main`](https://github.com/SergioBenitez/Rocket/blob/cc0621626bf5e917ea1d4c1e29a327e9a7871921/core/lib/src/lib.rs#L228) function. **However this function is private and should not be used explicitly** (see @SergioBenitez message below).

**Reply by the10thWiz** [reply]
@bubelov There are legitimate reasons why this is unsupported. The `async_main` function is unstable, and can be updated without warning, as long as the codegen is updated to match.

You could potentially avoid the issue by placing the launch code in a submodule, and calling the generated main function after parsing CLI args. However, I don't believe starting an async runtime is that intensive, and I would recommend testing to see if you can just start the runtime before parsing CLI args.

**Reply by bubelov** [reply]
@the10thWiz yea, that sounds like a path of least resistance. Thank you both for your suggestions!

**Reply by SergioBenitez** [reply]
> It is specified in the source code that the function is unstable and is not intended to be used outside Rocket. Otherwise, it is always possible to build yourself the tokio runtime man_shrugging

Please, please don't do this. This is a private function that is `pub` because Rust forces us to make it so. Don't ignore the docs:

```rust
/// WARNING: This is unstable! Do not use this method outside of Rocket!
```

Please don't suggest contradicting the docs.

**Reply by baptiste0928** [reply]
> Please don't suggest contradicting the docs.

Okay, sorry about that. I edited my answer to avoid confusion.

**Reply by cmcqueen** [reply]
Is it possible to use `#[rocket::main]` on a function called by `main()` and not on `main()` itself?

**Reply by the10thWiz** [reply]
> Is it possible to use `#[rocket::main]` on a function called by `main()` and not on `main()` itself?

Yes, however I believe the function generated by `rocket::main` is still called main. Placing it in a submodule allows you to use a different function as the actual main function, and call the generated function.

Something like this is the best option:
```rust
fn main() {
   // Check CLI Args, etc
   server::main();
}

mod server {
   #[launch]
   fn rocket() -> _ {
      rocket::build()
   }
}
```

If the rocket server isn't the main function of the program, it should probably be placed in a separate file anyway, which makes this type of change much simpler. Perhaps a parameter could be added to the main annotation to specify a different function name to generate, if this is intended to be a supported use case. Something like `#[rocket::main(name = "start_rocket")]` would make this easier to handle.

The other issue is that Rust only allows a single async runtime to be started. This means that this approach only works for running Rocket 0.5 from within sync code - meaning async Rust applications can't support this type of use case. A more complete solution would depend on Rust making async runtimes easier to swap out, which I believe is currently dependent on standardizing the `AsyncWrite` and `AsyncRead` traits. I don't expect this to happen anytime soon, and I don't see a way to support starting Rocket from within async code. Perhaps in the future Rocket could support running on an existing tokio runtime, but I don't think this is worthwhile yet.
